tests/client/test_async_client.py

```diff
@@ -146,3 +146,13 @@
 
     assert response.status_code == 200
     assert response.content == data
+
+
+async def test_async_client_timeout_default(server, backend):
+    async with httpx.AsyncClient(backend=backend) as client:
+        assert client.dispatch.timeout == httpx.config.UNSET
+
+
+async def test_async_client_timeout_disabled(server, backend):
+    async with httpx.AsyncClient(backend=backend, timeout=None) as client:
+        assert client.dispatch.timeout is None
```

Comment:
*Ideally*, to fully test tis we should make a request to the `/slow_response` endpoint of the server with a long delay (e.g. 100ms), surround the call with a "raise a timeout after x seconds" (where e.g. x = 50ms) instruction, and ensure that the timeout error is raised.

The "raise timeout after x seconds" instruction depends on the underlying async library, but we can add an util in `tests/concurrency.py` (single-dispatch stuff not listed here):

```python
import asyncio

async def timeout_after(delay, coroutine):
    try:
        await asyncio.wait(coroutine, timeout=delay)
    except asyncio.TimeoutError:
        raise TimeoutError


import trio

async def timeout_after(delay, coroutine):
    async with trio.move_on_after(delay) as cancel_scope:
        await coroutine
    if cancel_scope.cancelled_caught:
        raise TimeoutError
```

And use it like this:

```python
async def test_async_client_timeout_disabled(server, backend):
    url = server.url.replace_with(path="/slow_response/100")
    async with httpx.AsyncClient(backend=backend, timeout=None) as client:
        assert client.dispatch.timeout is None
        with pytest.raises(TimeoutError):
            await timeout_after(50e-3, client.get(url))
```

Or would this be too much?

Reply:
I was struggling a bit to change the default timeout of the client (didn't want to have a test that took the full 5s of the default timeout), but I really like this approach. I'll add a few tests with this pattern!